src/devices/cpu/drcbearm64.cpp

```diff
@@ -0,0 +1,3865 @@
+	assert_flags(inst, FLAG_V | FLAG_Z | FLAG_S);
+
+	be_parameter dstp(*this, inst.param(0), PTYPE_MR);
+	be_parameter edstp(*this, inst.param(1), PTYPE_MR);
+	be_parameter src1p(*this, inst.param(2), PTYPE_MRI);
+	be_parameter src2p(*this, inst.param(3), PTYPE_MRI);
+	const bool compute_hi = (dstp != edstp);
+
+	const a64::Gp src1 = src1p.select_register(TEMP_REG1, 8);
+	const a64::Gp src2 = src2p.select_register(TEMP_REG2, 8);
+	const a64::Gp lo = TEMP_REG3;
+	const a64::Gp hi = TEMP_REG2;
+
+	if ((src1p.is_immediate() && src1p.immediate() == 0) || (src1p.is_immediate() && src1p.immediate() == 0))
+	{
+		a.mov(lo, a64::xzr);
+		a.mov(hi, a64::xzr);
+	}
+	else
+	{
+		mov_reg_param(a, inst.size(), src1, src1p);
+		mov_reg_param(a, inst.size(), src2, src2p);
+
+		if (inst.size() == 8)
+		{
+			a.mul(lo, src1, src2);
+			a.smulh(hi, src1, src2);
+		}
+		else
+		{
+			a.smull(lo, src1.w(), src2.w());
+			a.lsr(hi, lo, 32);
+		}
+	}
+
+	mov_param_reg(a, inst.size(), dstp, lo);
+	if (compute_hi)
+		mov_param_reg(a, inst.size(), edstp, hi);
+
+	if (inst.flags())
+	{
+		a.mrs(SCRATCH_REG1, a64::Predicate::SysReg::kNZCV);
+
+		a.tst(lo, lo);
+		a.cset(TEMP_REG1, a64::CondCode::kEQ);
+		a.tst(hi, hi);
+		a.cset(SCRATCH_REG2, a64::CondCode::kEQ);
+		a.and_(TEMP_REG1, TEMP_REG1, SCRATCH_REG2);
+		a.bfi(SCRATCH_REG1, TEMP_REG1, 30, 1); // zero flag
+
+		if (inst.size() == 4)
+		{
+			a.sxtw(TEMP_REG1, lo.w());
+			a.cmp(TEMP_REG1, lo);
+		}
+		else
+		{
+			a.asr(TEMP_REG1, lo, 63);
+			a.cmp(TEMP_REG1, hi);
+		}
+
+		a.cset(TEMP_REG1, a64::CondCode::kNE);
+		a.bfi(SCRATCH_REG1, TEMP_REG1, 28, 1); // overflow flag
+
+		a.lsr(TEMP_REG1, hi, inst.size() * 8 - 1); // take top bit of result as sign flag
+		a.bfi(SCRATCH_REG1, TEMP_REG1, 31, 1); // sign flag
+
+		a.msr(a64::Predicate::SysReg::kNZCV, SCRATCH_REG1);
+	}
+}
+
+void drcbe_arm64::op_mulslw(a64::Assembler &a, const uml::instruction &inst)
+{
+	assert(inst.size() == 4 || inst.size() == 8);
+	assert_no_condition(inst);
+	assert_flags(inst, FLAG_V | FLAG_Z | FLAG_S);
+
+	be_parameter dstp(*this, inst.param(0), PTYPE_MR);
+	be_parameter src1p(*this, inst.param(1), PTYPE_MRI);
+	be_parameter src2p(*this, inst.param(2), PTYPE_MRI);
+
+	const a64::Gp src1 = src1p.select_register(TEMP_REG1, 8);
+	const a64::Gp src2 = src2p.select_register(TEMP_REG2, 8);
+	const a64::Gp lo = TEMP_REG3;
+	const a64::Gp hi = TEMP_REG2;
+
+	if ((src1p.is_immediate() && src1p.immediate() == 0) || (src1p.is_immediate() && src1p.immediate() == 0))
+	{
+		a.mov(lo, a64::xzr);
+		a.mov(hi, a64::xzr);
+	}
+	else
+	{
+		mov_reg_param(a, inst.size(), src1, src1p);
+		mov_reg_param(a, inst.size(), src2, src2p);
+
+		if (inst.size() == 8)
+		{
+			a.mul(lo, src1, src2);
+			a.smulh(hi, src1, src2);
+		}
+		else
+		{
+			a.smull(lo, src1.w(), src2.w());
+			a.lsr(hi, lo, 32);
+		}
+	}
+
+	mov_param_reg(a, inst.size(), dstp, lo);
+
+	if (inst.flags())
+	{
+		a.mrs(SCRATCH_REG1, a64::Predicate::SysReg::kNZCV);
+
+		a.tst(lo, lo);
```

Comment:
Same here. Sign-extending `lo` first when `inst.size() == 4` seems to make sense for this case.